### ISSUE TITLE
Provider Review Prep

### ISSUE THREAD

**Issue by cgriggs01** [AviatrixSystems/terraform-provider-aviatrix#2]
Hi Aviatrix Team,

We can use this thread to discuss and track the changes/additions that need to be addressed before the technical review can begin.

- [x] Source Code Repository: The repository must be organized with provider and resource files in an aviatirx/ directory while the main.go should be kept in the root or the repository. You can see the standard provider's repo organization here --> [Null_reasource](https://github.com/terraform-providers/terraform-provider-null)

- [x] Testing: Each resource in your provider must have acceptance testing that will be run nightly, once the provider is released.

- [x] Dependancies: All dependencies the provider uses should be vendored and checked into the source code --> [Null_resource vendor/](https://github.com/terraform-providers/terraform-provider-null/tree/master/vendor)

- [x] Documentation: Your provider and each resource must have corresponding documentation. Please follow the same format as other terraform providers documentation. Simple example --> [Null_reasource](https://github.com/terraform-providers/terraform-provider-null/tree/master/website)

- [x] LICENSE: Add a [license](https://github.com/terraform-providers/terraform-provider-null/blob/master/LICENSE) to the repository

- [x] Makefile: There is a [standard Makefile](https://github.com/terraform-providers/terraform-provider-null/blob/master/GNUmakefile) that is used for all providers.

- [x] CHANGELOG.md: Add a [changelog](https://github.com/terraform-providers/terraform-provider-null/blob/master/CHANGELOG.md) that will be used to version the provider. 

Let me know if you have any questions.

Best,
Chris

**Comment by rakesh568** [comment]
Chris,
I have addressed all the issues except Testing. Do you have an example/reference of acceptance testing that you mentioned.

Rakesh

**Comment by cgriggs01** [comment]
The best example that I could point you to would be the [AWS](https://github.com/terraform-providers/terraform-provider-aws) provider since it is maintained by our own Terraform engineers.

Thanks
Chris

**Comment by cgriggs01** [comment]
@rakesh568 I am just checking in checking in on your progress fo the provider's acceptance tests. 

Do you have any ETA for when they will be complete?

**Comment by AviatrixSystems** [comment]
Hi Chris, this is Sherry from Aviatrix. Rakesh left the company, I'm picking it up for now. Do you have more information on acceptance test in addition to the repo you pointed out? Much appreciated. -Sherry

**Comment by arvindsree** [comment]
Hello Chris,

Could you give more information on where the terraform tests will be run? This is because, we will need credentials to create cloud resources. Where are the credentials stored?

Thanks,
Arvind

**Comment by cgriggs01** [comment]
Hi Arvind,

Thanks for picking this up! The acceptance tests will eventually run nightly in our TeamCity CI workflow, but before that, myself and an engineer from the Terraform team will be running and testing the provider locally from our own machines. 

I hope this helps, let me know if you have any other questions.

Best,
Chris

**Comment by arvindsree** [comment]
This makes it clear. Will continue on getting the tests ready

**Comment by Deric4** [comment]
Not sure if this should be a separate issue or not but following a semantic versioning scheme for the provider would be nice. The provider installation documentation still references configuring `~/.terraformrc` which has been [deprecated](https://www.terraform.io/docs/commands/cli-config.html#deprecated-settings). 

Relavent Link: https://www.hashicorp.com/blog/hashicorp-terraform-provider-versioning

**Comment by arvindsree** [comment]
Hello Chris,

Acceptance tests should now be complete. How should we proceed from here

Thanks,
Arvind

**Comment by cgriggs01** [comment]
Hi @arvindsree,

Thanks for finishing those acceptance tests. The provider is currently in queue to be reviewed by a member of our Terraform Ecosystem Engineering team. Unfortunately they also have their hands full working to get Terraform 0.12 released ASAP. 

I will now close this issue and the reviewing engineer will open a new issue for their feedback. 

Let me know if you have any questions.

**Comment by arvindsree** [comment]
Hello Chris,

Could you let us know if you have any updates regarding the provider review.

@Deric4 : We have tried to version the provider using the Changelog.md and version file. Since we don't do binary releases currently, and the plugin binary is not renamed by the end user, terraform init will not show the version. However you can see the version in the Changelog.md file

Thanks,
Arvind

**Comment by arvindsree** [comment]
Hello Chris,

Do you have any updates regarding the provider acceptance?

Thanks,
Arvind

**Comment by cgriggs01** [comment]
Hi Team, 

My apologies, I seemed to have missed the updates from this issue thread.

We are starting the second review pass now, with the changes coming from the newly released Terraform 0.12, I'd suggest reading through our [blog post](https://www.hashicorp.com/blog/announcing-terraform-0-12?mkt_tok=eyJpIjoiTVdReU5qUTFNamt6TjJVeSIsInQiOiJwNzhpMHdON0M5ZmZNQzN4RGZiS1UreDRrc2xDc3F3SEVndzVIaG1cLzZlOXJFZWJqYXRzNUxOYjQzTVlWMGVUZzhlRmx5RG81VVVwSzY2Qk90cGJPTXE5ZU9NVEJxSU9rdFFnRzRFaFBMa2FSSjMwWUk5NkRiNmdSaEhNTTNBKzQifQ%3D%3D) and plan to upgrade to the new Terraform 0.12 SDK before releasing this provider officially. 

- [ ] With the upgrade to Go v1.11.5 or greater, all Terraform provider are now using Go Modules for managing dependancies. As well there is still a requirement pull all dependancies to a vendor/ directory within the repository. 

- [ ] There naming convention for resource file should contain the name of the providers. i.e. `resource_account.go` should be `resource_aviatix_account.go`

- [ ] Once the provider is released, we'll run the acceptance tests nightly on a TeamCity server. It looks we need to deploy and configure an Aviatix environment on AWS. Could you provide HashiCorp with a BYOL to use when deploying the Aviatrix AMI. If so, please submit the credentials [development form](https://docs.google.com/forms/d/e/1FAIpQLSe5TjsThekgY0nUZan8M8JFVFk2qFC0shR6Fwpggn8XhhoHdg/viewform?usp=sf_link) or email at terraform-acctest@hashicorp.com. We have some AWS credits to use for this testing, so all we'll need is Aviatrix credentials. 

- [ ]  Throughout your code base their are a number of lines of code that are commented out and unused. These commented should be removed before moving forward with the release. 

- [ ] Within the resource schema, it's common practice to group the `Required` arguments at the top of the schema, then followed by `Optional` arguments and finally the computed attributes.

- [ ] If a resource cannot be found(404) in the read function, we should do a d.SetId("") to remove it from the statefile. You also wouldn't need to return an error then. Terraform will just treat it as a resource that never existed and try and create it.

- [ ] In the resource [AdminEmail](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/aviatrix/resource_admin_email.go#L63) and [CustomerID](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/aviatrix/resource_customer_id.go#L66) `DELETE` method has not been implemented. 

- [ ] The CREATE method of any Resource should at least return the results of it’s own READ method. In certain cases it may need to return the results of the UPDATE method, depending on if the CREATE method is capable of setting all possible configuration values. If for example the API call to create Resource only allows basic creation, the remaining configuration is applied via update, then CREATE should return UPDATE, and UPDATE should return READ. [AWS Security Group Resource](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_security_group.go) is an example of this; the CREATE only allows creating a Security Group with a name and description, and we must make other API calls to apply the rules. Thus, CREATE returns the results of UPDATE. 

- [ ] All the attributes should be read and set on Read so that differences between the configuration and the actual state can be caught. Any non-scalar values (TypeList, TypeSet, TypeMap) should have errors checked when setting.

 - [ ] If a resource doesn't not have any update functionality, it should be completely removed from the resource file.

 - [ ] In your website documentation the sidebar ruby layout file [aviatrix.erb](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/website/aviatrix.erb) is missing most of the links to navigate to the different resource and data source doc pages. 

- [ ] The documentation file [aviatrix_vpc_html.markdown](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/website/docs/r/aviatrix_vpc_html.markdown) name needs to be updated to `aviatrix_vpc.html.markdown`

Please let me know if you have any questions.

**Comment by arvindsree** [comment]
Hello Chris,

We have sent the BYOL license id using the google form and will get the remaining items completed soon.

As of now the provider is not compatible with terraform 0.12 . Would we have to drop support for terraform .011 and publish the provider only after making it compatible with 0.12? Also is terraform 0.12 officially released or still in rc.

Regarding (1)The dependencies have been vendored, we shall integrate go mod as well
Regarding (7) the resource AdminEmail and CustomerID DELETE are mainly configurations and are not resources in the terraform sense, and are needed to be set only once. Customer ID, if set to empty on deletion would cause the remainder of the acceptance tests to fail. Should we just deprecate them completely? or move them to the ControllerConfig resource?

Also, How would github codebase be maintained after the acceptance is complete? Would it be moved to github.com/terraform-providers or remain in this same github repo? ( The fix for issue #97 will depend on that) 


Thanks,
Arvind

**Comment by zhiqiangjin** [comment]
Hello Chris,

I am a little confused about the 9th you listed above:

"All the attributes should be read and set on Read so that differences between the configuration and the actual state can be caught. Any non-scalar values (TypeList, TypeSet, TypeMap) should have errors checked when setting."

Do you mean we can not set any attributes in the CREATE function? Like our aviatrix_gateway resource, after creating gateway in CREATE function, we also need to do something else, and we probably need to set some attributes just in case that it fails and no state is set. If you can give an example, or explain more, it will be great.

Thanks,
Zhiqiang Jin

**Comment by cgriggs01** [comment]
Hi @arvindsree, Sorry for the late reply. Now that Terraform 0.12 GA has now been released, we're now requiring that all new provider upgrade to the Terraform 0.12 SDK. 

Once the provider has been certified, we will clone this repo over to a newly created repo at github.com/terraform-providers/terraform-provider-aviatrix/, then you and your team will be granted "write" access over that new repo. 

@zhiqiangjin Yes, this is why the Create and Update should return their own Read method, when the Read method is used to write and changes to the Terraform state. 

The error checking is required on any complex attribute being called with d.Set(). The instance I was referring to was in the [Firewall resource](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/aviatrix/resource_aviatrix_firewall.go#L178) that should be handled like the [AWS Security Group Resource](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/aviatrix/resource_aviatrix_firewall.go#L178) 

I hope that helps. Let me know if you have any other question.  

The only other thing we need to figure out is the infrastructure environment needed for the testing of this providers. I can provider credentials for deploying the required AWS|GCP|Azure infra for testing. But we need to write a TF configuration for said infra to deploy each night and run this providers acceptance tests against. Do you have any thoughts on this? 

Thanks,
Chris.

**Comment by arvindsree** [comment]
Hello Chris,

Thanks for the reply. It is possible to set up the acceptance test environment using a tf file. We will get back on the tf file to create the acceptance test environment. It would basically set up the VPCs/Vnets/Controller as per [Readme]( https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/README_accep_test.md). 

Regarding @zhiqiangjin's question, we feel that it is better to do partial d.Sets rather than reading the refreshing the entire resource, in the case where a partial operation to modify an attribute(after the resource is created) fails. I guess this will have to be done by a "defer Readfunction()" which would make the create and update a bit slower in general.

Thanks,
Arvind

**Comment by cgriggs01** [comment]
Hey @arvindsree,

Yes, a TF config file that deploys everything needed to test the providers would be perfect. This config should be kept in a `test-infra/` directory in this repo.

**Comment by arvindsree** [comment]
Hello Chris,

All the points you have mentioned should be resolved now. The test-infra now contains the terraform files required to build the infrastructure required to run the acceptance tests. (outputs.tf will output the required environment variables). 

Thanks,
Arvind

**Comment by walkafwalka** [comment]
@cgriggs01, as an Aviatrix client, I am really looking forward to this.

**Comment by cgriggs01** [comment]
Hey Team, apologies for the wait. I was away on leave but back and ready to move forward with the finalization and release of this provider. 

The next step is to upload the provider code to the github.com/terraform-providers/terraform-provider-aviatrix repository and great whomever on your team needs "write" access of the repository. Could you provider a list of maintainers for us to grant privileges to?

**Comment by zhiqiangjin** [comment]
Hi Chris,

The maintainers are:

zhiqiangjin
arvindsree
iggynopolis
inc1t3Ful
edumanig

Thanks,
Zhiqiang Jin
